### PR TITLE
fix(ec2): export NatGatewayProvider for consistency with NatInstanceProvider

### LINKED ISSUE #28372: aws-ec2: Export NatGatewayProvider
### Describe the feature

(originally posted as [a discussion](https://github.com/aws/aws-cdk/discussions/28154))

There are two implementations of [NatProvider](https://github.com/aws/aws-cdk/blob/c6471f2f603e2570e8781d6e1ae7d8252aec0f79/packages/aws-cdk-lib/aws-ec2/lib/nat.ts#L55-L100): [NatGatewayProvider](https://github.com/aws/aws-cdk/blob/c6471f2f603e2570e8781d6e1ae7d8252aec0f79/packages/aws-cdk-lib/aws-ec2/lib/nat.ts#L217-L261) and [NatInstanceProvider](https://github.com/aws/aws-cdk/blob/c6471f2f603e2570e8781d6e1ae7d8252aec0f79/packages/aws-cdk-lib/aws-ec2/lib/nat.ts#L266-L356).

Of the two, only NatInstanceProvider is exported. Both are usable through `aws_ec2.NatProvider.instance()` and `aws_ec2.NatProvider.gateway()` respectively, but since only NatInstanceProvider is exported [the docs](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_ec2.NatProvider.html) indicate that `NatProvider` is only implemented by `NatInstanceProvider`.

I suggest that NatGatewayProvider also be exported, so that both providers are available through both interfaces and so the docs correctly reflect the two available implementations of `NatProvider`.

### Use Case

No real new functionality is exposed, except the ability to instantiate NatGatewayProvider directly instead of using the static method of NatProvider. The main advantage is the implied update to the docs.

### Proposed Solution

I think this is as simple as inserting an `export` at the start of L217 https://github.com/aws/aws-cdk/blob/c6471f2f603e2570e8781d6e1ae7d8252aec0f79/packages/aws-cdk-lib/aws-ec2/lib/nat.ts#L214-L261

### Other Information

_No response_

### Acknowledgements

- [X] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### CDK version used

2.115.0

### Environment details (OS name and version, etc.)

Windows 10

### PR DESCRIPTION
**Pull request by crh23**
Export `NatGatewayProvider` for two reasons: to allow instantiation with `new`, and to make docs present it as an implementation of `NatProvider`.
Also added a unit test for the same "functionality" for `NatInstanceProvider`, for symmetry with the added test for `NatGatewayProvider`. 
Unit test and integration test for new functionality added.

Closes #28372.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by crh23** [comment]
Clarification Request: this change is somewhere between a feature and a bug fix ([actual change is tiny](https://github.com/aws/aws-cdk/pull/28810/files#diff-1228e5eb29751308416d9af171ef17872ef911c3b2f1249a78e120a1f22c3565)).

1. Should I relabel it as a bug fix?
2. If not, is a README change required? The enabled functionality is very minor, and the previously-available version isn't in the README currently (closest is "[Using NAT instances](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_ec2-readme.html#using-nat-instances)")

**Comment by crh23** [comment]
Applying a bit more thought, I think this is a bug fix more than a feature - the change to actual functionality is the addition of a single `export`, and it seems unintentional that that wasn't already present,

File: packages/aws-cdk-lib/aws-ec2/test/vpc.test.ts
Comment (msambol) [inline packages/aws-cdk-lib/aws-ec2/test/vpc.test.ts:1654]: `Can instantiate NatInstanceProvider directly with new`

File: packages/aws-cdk-lib/aws-ec2/test/vpc.test.ts
Comment (msambol) [inline packages/aws-cdk-lib/aws-ec2/test/vpc.test.ts:1261]: nit: `Can instantiate NatGatewayProvider directly with new`

File: packages/@aws-cdk-testing/framework-integ/test/aws-ec2/test/integ.nat-gateway-provider.ts
Comment (msambol) [inline packages/@aws-cdk-testing/framework-integ/test/aws-ec2/test/integ.nat-gateway-provider.ts:9]: what's `/// !show` ?

**Review by msambol — CHANGES_REQUESTED** [review]
Few nits but looks good to me.

File: packages/@aws-cdk-testing/framework-integ/test/aws-ec2/test/integ.nat-gateway-provider.ts
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk-testing/framework-integ/test/aws-ec2/test/integ.nat-gateway-provider.ts:9]: We don't use pragma anymore.

File: packages/@aws-cdk-testing/framework-integ/test/aws-ec2/test/integ.nat-gateway-provider.ts
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk-testing/framework-integ/test/aws-ec2/test/integ.nat-gateway-provider.ts:27]: What are these for?

**Comment by TheRealAmazonKendra** [comment]
Hmm, not totally sure how I feel about this one, not because anything you've done here is invalid but because it looks to me like the error here was exporting the other function, not the lack of export on this one. We can't undo the export because that would be a breaking change, but I'm not sure if it's better to have it consistently wrong in both places or inconsistent.

**Comment by crh23** [comment]
> Hmm, not totally sure how I feel about this one, not because anything you've done here is invalid but because it looks to me like the error here was exporting the other function, not the lack of export on this one. We can't undo the export because that would be a breaking change, but I'm not sure if it's better to have it consistently wrong in both places or inconsistent.

I've seen folks be confused by the current inconsistency. In the design of the CDK docs, the natural way (for me) to find the implementations of an abstract class (such as `NatProvider`) is to view the classes listed in _Implemented by_. Currently this misses `NatGatewayProvider`, but includes `NatInstanceProvider`. The new `NatInstanceProviderV2`, which was added since I raised this CR, is also `export`ed.

I think I agree that it would be best to `export` none of them (and add a usage note to the docs of `NatProvider`), but that's not an option now, and I think `export`ing one but not the other (current situation) is worse than `export`ing all (this CR).

**Review by TheRealAmazonKendra — APPROVED** [review]
> I've seen folks be confused by the current inconsistency. In the design of the CDK docs, the natural way (for me) to find the implementations of an abstract class (such as NatProvider) is to view the classes listed in Implemented by. Currently this misses NatGatewayProvider, but includes NatInstanceProvider. The new NatInstanceProviderV2, which was added since I raised this CR, is also exported.
>
> I think I agree that it would be best to export none of them (and add a usage note to the docs of NatProvider), but that's not an option now, and I think exporting one but not the other (current situation) is worse than exporting all (this CR).

Fair point. 

YOLO

**Review by TheRealAmazonKendra — DISMISSED** [review]

**Comment by TheRealAmazonKendra** [comment]
Drat. This failing test has nothing to do with your change and everything to do with how long your waiting on this approval. I'll fix it.

**Comment by crh23** [comment]
> Drat. This failing test has nothing to do with your change and everything to do with how long your waiting on this approval. I'll fix it.

Ack - I recall getting integ tests working in my environment was a bit of a chore, so happy to let you do it this time 😁 
(assuming that is what the issue is)

**Review by TheRealAmazonKendra — APPROVED** [review]
YOLO Again. I should probably stop saying that in official PR reviews... ¯\_(ツ)_/¯

**Pull request merged**
